src/components/tl/ucp/tl_ucp_context.c

```diff
@@ -0,0 +1,126 @@
+    ucp_worker_attr_t   worker_attr;
+    ucp_params_t        ucp_params;
+    ucp_config_t       *ucp_config;
+    ucp_context_h       ucp_context;
+    ucp_worker_h        ucp_worker;
+    ucs_status_t        status;
+
+    UCC_CLASS_CALL_SUPER_INIT(ucc_tl_context_t, tl_config->tl_lib);
+    status = ucp_config_read(params->prefix, NULL, &ucp_config);
+    if (UCS_OK != status) {
+        tl_error(&self->super, "failed to read ucp configuration, %s",
+                 ucs_status_string(status));
+        ucc_status = ucs_status_to_ucc_status(status);
+        goto err_cfg;
+    }
+
+    ucp_params.field_mask =
+        UCP_PARAM_FIELD_FEATURES | UCP_PARAM_FIELD_REQUEST_SIZE |
+        UCP_PARAM_FIELD_REQUEST_INIT | UCP_PARAM_FIELD_REQUEST_CLEANUP |
+        UCP_PARAM_FIELD_TAG_SENDER_MASK;
+    ucp_params.features        = UCP_FEATURE_TAG | UCP_FEATURE_RMA;
+    ucp_params.request_size    = sizeof(ucc_tl_ucp_req_t);
+    ucp_params.request_init    = ucc_tl_ucp_req_init;
+    ucp_params.request_cleanup = ucc_tl_ucp_req_cleanup;
+    ucp_params.tag_sender_mask = UCC_TL_UCP_TAG_SENDER_MASK;
+
+    if (params->estimated_num_ppn > 0) {
+        ucp_params.field_mask |= UCP_PARAM_FIELD_ESTIMATED_NUM_PPN;
+        ucp_params.estimated_num_ppn = params->estimated_num_ppn;
+    }
+
+    if (params->estimated_num_eps > 0) {
+        ucp_params.field_mask |= UCP_PARAM_FIELD_ESTIMATED_NUM_EPS;
+        ucp_params.estimated_num_eps = params->estimated_num_eps;
+    }
+
+    status = ucp_init(&ucp_params, ucp_config, &ucp_context);
+    ucp_config_release(ucp_config);
+    if (UCS_OK != status) {
+        tl_error(&self->super, "failed to init ucp context, %s",
+                 ucs_status_string(status));
+        ucc_status = ucs_status_to_ucc_status(status);
+        goto err_cfg;
+    }
+
+    worker_params.field_mask = UCP_WORKER_PARAM_FIELD_THREAD_MODE;
+    switch (params->thread_mode) {
+    case UCC_THREAD_SINGLE:
+    case UCC_THREAD_FUNNELED:
+        worker_params.thread_mode = UCS_THREAD_MODE_SINGLE;
+        break;
+    case UCC_THREAD_MULTIPLE:
+        worker_params.thread_mode = UCS_THREAD_MODE_MULTI;
+        break;
+    default:
+        /* unreachable */
+        ucc_assert(0);
+        break;
+    }
+    status = ucp_worker_create(ucp_context, &worker_params, &ucp_worker);
+    if (UCS_OK != status) {
+        tl_error(&self->super, "failed to create ucp worker, %s",
+                 ucs_status_string(status));
+        ucc_status = ucs_status_to_ucc_status(status);
+        goto err_worker_create;
+    }
+
+    if (params->thread_mode == UCC_THREAD_MULTIPLE) {
+        worker_attr.field_mask = UCP_WORKER_ATTR_FIELD_THREAD_MODE;
+        ucp_worker_query(ucp_worker, &worker_attr);
+        if (worker_attr.thread_mode != UCS_THREAD_MODE_MULTI) {
+            tl_error(&self->super,
+                     "thread mode multiple is not supported by ucp worker");
+            status = UCC_ERR_NOT_SUPPORTED;
+            goto err_thread_mode;
```

Comment:
From the UCC perspective, it is ok to return single thread mode, even if multithreaded was requested.

Reply:
@manjugv  This one is tricky. UCC defines THREAD mode as LIB object attribute. So, all the checks by user are done before creating the context. In contrast, ucp threading mode is defined (queried) on the WORKER, which is only created during context creation. So, if at this point i will return OK, saying to upper layer (CL) that i can only support THREAD_SINGLE (even if ucc_lib is MT) we can no longer change the level of threading support that has been already reported to user via lib_attr. 

Thoughts ? Does this mean that we need to move UCC_THREAD_MODE to the context ?

Reply:
If the library is created in single-thread mode - all teams and contexts are in single-threaded mode. If the library is multithreaded, then the context can be either single or multi-threaded. I think we need to expose thread mode in context, so we have finer control.

Reply:
After discussion we decided it is not right to change UCC iface only because UCP is implemented that way. Ideally we should be able to query UCP if it can support THREAD_MULTIPLE w/o allocation of ucp_worker. Then the issue above will be resolved. So far we keep the error handling that way: reporting error if TL can't support required thread level